lib/pka.c

```diff
@@ -1694,6 +1748,166 @@ int pka_modular_inverse(pka_handle_t   handle,
     return pka_submit_cmd(handle, user_data, CC_MODULAR_INVERT, &operands);
 }
 
+int pka_is_mont_ecdh_canonical(ecc_mont_curve_t* curve,
+                               uint16_t          curve_type,
+                               pka_operand_t*    point_x)
+{
+    pka_comparison_t rc;
+    uint32_t         idx;
+    uint8_t          big_endian, ls_byte, ms_byte;
+
+    big_endian = point_x->big_endian;
+    if (curve_type == PKA_CURVE_25519)
+    {
+        if (point_x->actual_len != 32)
+            return 1;
+
+        // We want to see if point_x (with a special adjustment of the
+        // most significant byte) is < the curve prime.
+        // First do a quick test
+        ls_byte = point_x->buf_ptr[big_endian ? 31 : 0];
+        if (ls_byte < 0xED) // 237 = 256 - 19
+            return 1;
+
+        // Loop over the bytes from least significant to most significant
+        // looking for a byte != 0xFF.  The most signifcant byte is special.
+        if (big_endian)
+        {
+            for (idx = 1; idx <= 30; idx++)
+                if (point_x->buf_ptr[31 - idx] != 0xFF)
+                    return 1;
+
+            ms_byte = point_x->buf_ptr[0];
+        }
+        else
+        {
+            for (idx = 1; idx <= 30; idx++)
+                if (point_x->buf_ptr[idx] != 0xFF)
+                    return 1;
+
+            ms_byte = point_x->buf_ptr[31];
+        }
+
+        ms_byte &= 0x7F;
+        return ms_byte != 0x7F;
+    }
+    else if (curve_type == PKA_CURVE_448)
+    {
+        if (point_x->actual_len != 56)
+            return 1;
+
+        // Quick test *TBD*
+        ms_byte = point_x->buf_ptr[big_endian ? 0 : 55];
+        if (ms_byte != 0xFF)
+            return 1;
+
+        rc = pka_internal_compare(point_x->buf_ptr, curve->p.buf_ptr, 56,
+                                  big_endian);
+        if (rc == PKA_LESS_THAN)
+            return 1;
+
+        return 0;
+    }
+    else
+        return 1;
+}
+
+int pka_mont_ecdh_canonicalize(pka_handle_t      handle,

```

Comment:
If we have defined "is_canonical" flag we don't need to call pka_is_mont_ecdh_canonical()

By the way, I would have thought of this function to be implemented by the test framework... Not very sure about that, though. The idea is to abstract the curve information from the API and let the test or any other application to pass the valid input (with all the modification and adjustments). Perhaps, OpenSSL implementation would be a good reference a give us a bit of guidance here... Do you know whether the parameters are passed in the correct format or whether they need sort of adjustment after the API is called?

Reply:
Maybe let's keep it in the API, but let's not expose it. Should be an internal processing.